engine/servermaster/config.go

```diff
@@ -154,74 +112,47 @@ func (c *Config) Toml() (string, error) {
 	return b.String(), nil
 }
 
-// Parse parses flag definitions from the argument list.
-func (c *Config) Parse(arguments []string) error {
-	// Parse first to get config file.
-	err := c.flagSet.Parse(arguments)
+// Clone clones a master config
+func (c *Config) Clone() *Config {
+	data, err := json.Marshal(c)
 	if err != nil {
-		return errors.WrapError(errors.ErrMasterConfigParseFlagSet, err)
-	}
-
-	if c.printVersion {
-		fmt.Print(version.GetRawInfo())
-		return flag.ErrHelp
-	}
-
-	// Load config file if specified.
-	if c.ConfigFile != "" {
-		err = c.configFromFile(c.ConfigFile)
-		if err != nil {
-			return err
-		}
+		log.Panic("failed to marshal config", zap.Error(err))
 	}
-
-	// Parse again to replace with command line options.
-	err = c.flagSet.Parse(arguments)
+	clone := new(Config)
+	err = json.Unmarshal(data, &clone)
 	if err != nil {
-		return errors.WrapError(errors.ErrMasterConfigParseFlagSet, err)
-	}
-
-	if len(c.flagSet.Args()) != 0 {
-		return errors.ErrMasterConfigInvalidFlag.GenWithStackByArgs(c.flagSet.Arg(0))
+		log.Panic("failed to unmarshal config", zap.Error(err))
 	}
-	return c.adjust()
+	return clone
 }
 
-func (c *Config) adjust() (err error) {
+// Adjust adjusts the master configuration
+func (c *Config) Adjust() (err error) {
 	c.Etcd.Adjust(defaultPeerUrls, defaultInitialClusterState)
 
 	if c.AdvertiseAddr == "" {
 		c.AdvertiseAddr = c.MasterAddr
 	}
 
-	if c.KeepAliveIntervalStr == "" {
-		c.KeepAliveIntervalStr = defaultKeepAliveInterval
-	}
 	c.KeepAliveInterval, err = time.ParseDuration(c.KeepAliveIntervalStr)
 	if err != nil {
 		return err
 	}
 
-	if c.KeepAliveTTLStr == "" {
-		c.KeepAliveTTLStr = defaultKeepAliveTTL
-	}
 	c.KeepAliveTTL, err = time.ParseDuration(c.KeepAliveTTLStr)
 	if err != nil {
 		return err
 	}
 
-	if c.RPCTimeoutStr == "" {
-		c.RPCTimeoutStr = defaultRPCTimeout
-	}
 	c.RPCTimeout, err = time.ParseDuration(c.RPCTimeoutStr)
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
-// configFromFile loads config from file.
-func (c *Config) configFromFile(path string) error {
+// ConfigFromFile loads config from file.
+func (c *Config) ConfigFromFile(path string) error {
```

Comment:
How about `func LoadConfig(path string) (*Config, error)`? I think it's easier to use.

Reply:
This function also merges configurations into Config, and only used in test. I have added some comments about it.